### PR TITLE
Upgrade to v4.0

### LINKED ISSUE #262: v4.0
This is an issue tracking this library's upgrade to v4.0.

New features:
- [x] Support for Bulma 0.9.1
    - [x] Spacing helpers 
    - [x] Light/Dark color helpers
    - [x] Remove `<List />` component
    - [x] `is-clickable` helper
    - [x] `is-max-desktop` and `is-max-widescreen` for Container
- [x] Standalone `display` prop without viewport modifiers. (5f1b1eb48e53ac87ac95ddcd53c3e7922323db98)
- [x] Support `is-size-7` (2162f3fb4d38f31a4c52446860c12fa098fd3161)
- [x] Support font family modifiers (d82cee85e89d391e6eae18ff703032684be30589)
- [x] Support for `is-sr-only` modifier (e9fc3d2854298988b1b8136619688a3a86fcc960)
- [x] Tree shaking
- [x] Update Documentation for installation process (Now there will be only one no more advanced configuration need it)
- [x] TypeScript support
- [x] Add missing `table-container` component (#274)
- [x] Add missing `hoverable` prop for `Table` component (#273)
- [x] Add missing `rounded` prop for `Form.Select` component (aed577435c2f6070d49e1d53102964232beb4a21) 
- [x] Add various state props for `Form.Select` (d2afa6dcc4f2b3de383db7ac2ba4cc1120934bd4)
- [x] **BREAKING** `name` prop of `Form.Radio` is no longer required. (b365663a0bdd8ff3bb2ecf6fae4722cb09c4335e)
- [x] **BREAKING** Simplify responsive props (#279)
- [x] **BREAKING** remove `Breadcrumb` `item` prop and replace it with `Breadcrumb.Item`
- [x] Ability to show the first and last page in Pagination (#283)
- [x] **BREAKING** ~Remove `loading` Button prop in favor of a unified `state` prop that already exists.~ Button can have multiple state at once - the `state` prop will be removed because it can only take one state at a time. (0da86324c10f9182c9904149e0d1422ea1b38bf6) **Reverting this, does not make sense to have focus, hover and/or hover at the same time**
- [x] Add focus, hover ~and loading~ (**UPDATE** loading state is applied to Control instead of Input, removing) state props for `Form.Input`. (aff9690ef622413c93a92aeafca21ad72f659523)
- [x] Add `rounded` prop for `Form.Input` (corresponds to `is-rounded`) (af1c3a2a063db32eb6b0f973ca666ce0d2910b4b).
- [x] Add `centered` prop for `InputFile`
- [x] **BREAKING** Remove `Form.Textarea` default rows of 4. (2153d5b20f1a763c12621096d50ff88e8caf2357) 
- [x] Add `has-fixed-size` support for `Form.Textarea` (2153d5b20f1a763c12621096d50ff88e8caf2357)
- [x] Add various state props (loading, focused, etc.) for `Form.Textarea` (2153d5b20f1a763c12621096d50ff88e8caf2357) **Reverting this, does not make sense to have focus, hover and/or hover at the same time**
- [x] **BREAKING** `Form.InputFile` is now a controlled component
- [x] ~**BREAKING** `Heading` will now pick the appropriate header element based on the supplied size. For example, `size={4}` tells `Heading` to render `h4` under the hood. Default size will now be 3~ **Won't do, the user will select the appropiate element with renderAs**
- [x] Allow override of `renderAs` for various components for consistency's sake. **Except Table because with any other component will not work**

Housekeeping:
- [x] **BREAKING** Move `bulma` to `peerDependency` (#258)
- [x] **BREAKING** Move Remove sass import from codebase
- [x] Remove yarn.lock
- [x] (Optional) Rewrite class components with hooks.
- [x] (Optional) Review how we exports the components and see if there are no a better way
- [x] Remove default HTML attributes/React props from propTypes object (for example style, disabled) unless it is accessed by the component.

cc @couds 

edit: marked breaking changes.

### PR DESCRIPTION
**Pull request by kennethnym**
This PR tracks the upgrade of this library to v4.0

Closes #262

### CODE REVIEW

**Comment by MrCreeper1008** [comment]
@couds is this typescript structure good?

**Comment by couds** [comment]
> @couds is this typescript structure good?

wdyt about moving the definitions that still are inside the `src/index.d.ts` to for example `src/helpers/index.d.ts`? I would prefer this `src/index.d.ts` only importing/exporting the definitions. that make sense?

**Comment by MrCreeper1008** [comment]
yeah that sounds alright to me.

**Comment by MrCreeper1008** [comment]
@couds need to rewrite all stories since storybook v6 is a significant api change from v5.

**Comment by MrCreeper1008** [comment]
@couds v4 should be largely complete.

**Comment by fgs-dbudwin** [comment]
Very excited about this PR...just wanted to chime in and ask if there was an ETA for this? 👍

**Comment by MrCreeper1008** [comment]
@fgs-dbudwin i am waiting for @couds's review.

**Comment by MrCreeper1008** [comment]
@couds can we do a beta release of v4 while we iron out all the bugs?

**Comment by akelsch** [comment]
@MrCreeper1008 any update on this? I would love to make use of the new spacing and flexbox helpers in Bulma!

**Comment by MrCreeper1008** [comment]
I'm waiting on @couds to do a beta release :/ I can't release the library on non

**Comment by dbudwin** [comment]
I noticed the `package.json` is pinning Bulma to version 0.9.2.  Maybe I'm missing something, but it appears the latest release is 0.9.1 so I get issues trying to install the `next` branch.  See: https://github.com/jgthms/bulma/issues/3212

**Comment by MrCreeper1008** [comment]
Sorry about that. Will fix.

**Comment by MrCreeper1008** [comment]
Can't upgrade storybook due to this issue: https://github.com/babel/babel/issues/11622. 
Every workaround in the thread **does not work**.

**Comment by couds** [comment]
@MrCreeper1008 I just released a new pre-release version on @next

**Comment by MrCreeper1008** [comment]
Nice! Thank you so much.

File: src/components/navbar/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { BulmaComponent } from '..';
+import { Color } from '..';
+
+interface NavbarProps {
+  transparent?: boolean;
+  fixed?: 'top' | 'bottom';
+  size: 'large';
```
Comment (davepwsmith) [inline src/components/navbar/index.d.ts:7]: Is this supposed to be a mandatory prop? Seems especially odd since there's only one allowed value.

Comment (couds) [reply src/components/navbar/index.d.ts:7]: you are right. this is an optional prop.

**Comment by kennethnym** [comment]
@couds that was quick! was gonna start working on it. thanks!

**Pull request merged**